[Performance] Implement command loader to reduce initialization time of Magento CLI

mattwellss opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
This change reduces initialization time of the Magento CLI app by allowing deferred initialization of commands until they're needed.

* Implements the Symfony CommandLoaderInterface three times:
  1. `\Magento\Setup\Console\CommandLoader`, a replacement for `Magento_Setup`'s `CommandList`
  1. `\Magento\Framework\Console\CommandLoader`, an alternative means of adding commands outside the `Magento\Setup` namespace
  1. `\Magento\Framework\Console\CommandLoader\Aggregate`, an aggregate command loader, necessary because a Symfony CLI app can only have one Command Loader.

### Related Pull Requests
<!-- related pull request placeholder -->
None

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#29266

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Check out the 2.4-develop branch
1. Run a console command through `time`; for exmaple: `time ./bin/magento setup:install --help` and note how long it takes
1. Check out this branch locally (mattwellss:implement-command-loader)
2. Run the same console command again, and note that it is significantly faster than before

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

* Should the "Aggregate" command loader be renamed or implemented a different way? (Extending `\Magento\Framework\ObjectManager\TMap` seems feasible)
* I identified `setup:install` and `setup:config:set` as the two commands causing the most slowdown, so I focused first on migrating the commands in `Magento_setup`. I'm happy to migrate more commands to use the Command Loader, but it seems unnecessary from a performance perspective.
* Should `Magento\Setup\Console\CommandList` be removed entirely, or just have its `getCommandsClasses()` method updated to return an empty array? The latter would preserve BC, in the case that people have extended the class with their own commands.


### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

lbajsarowicz commented on lib/internal/Magento/Framework/App/Test/Unit/Console/CommandLoader/AggregateTest.php:
```diff
@@ -0,0 +1,90 @@
+<?php
+
```
1. Add missing license
2. Introduce strict types

lbajsarowicz commented on lib/internal/Magento/Framework/App/Test/Unit/Console/CommandLoader/AggregateTest.php:
```diff
@@ -0,0 +1,90 @@
+
+class AggregateTest extends TestCase
+{
+    private $firstCommandLoader;
+    private $secondCommandLoader;
+    private $subj;
```
1. Add missing annotations
2. Do not use shortcuts for property names.

lbajsarowicz commented on lib/internal/Magento/Framework/App/Test/Unit/Console/CommandLoader/AggregateTest.php:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * @dataProvider provideTestCasesForHas
+     */
+    public function testHas($firstResult, $secondResult, $overallResult)
```
The test method name should be more descriptive about what it tests, what is the expected outcome.

mattwellss replied:
I spent some time thinking about how to make this name better, and I'm not sure what else to add. Because it uses a data provider, it's meant to cover every known case of the method.

I could switch strategies and build multiple test methods: `testHasWithAtLeastOneMatch`, `testHasWithZeroMatches`, etc. but I'm not sure that's better (aside from having descriptive names). I did add significant prose to the docblock, which I believe helps explain what the test achieves.

lbajsarowicz commented on lib/internal/Magento/Framework/Console/CommandLoader/Aggregate.php:
```diff
@@ -0,0 +1,66 @@
+     * @param string $name
+     * @return \Symfony\Component\Console\Command\Command
+     */
+    public function get($name)
```
Introduce strict type
```suggestion
    public function get(string $name)
```

mattwellss replied:
I don't believe this is possible–the [interface being implemented does not use strict types](https://github.com/symfony/console/blob/4.4/CommandLoader/CommandLoaderInterface.php#L31).

Here's an example in action: https://3v4l.org/gml83

ihor-sviziev replied:
But seems like return type we still could define
```suggestion
    public function get($name): \Symfony\Component\Console\Command\Command
```

Here is an example in action: https://3v4l.org/nQTa0

lbajsarowicz commented on lib/internal/Magento/Framework/App/Test/Unit/Console/CommandLoader/AggregateTest.php:
```diff
@@ -0,0 +1,90 @@
+     */
+    public function testHas($firstResult, $secondResult, $overallResult)
+    {
+        $this->firstCommandLoader->method('has')->with('foo')->willReturn($firstResult);
+        $this->secondCommandLoader->method('has')->with('foo')->willReturn($secondResult);
+
+        static::assertEquals($overallResult, $this->subj->has('foo'));
```
In Magento we generally use `$this->`. It's not a written standard, but for sake of consistency, I'd recommend using the same.

lbajsarowicz commented on lib/internal/Magento/Framework/App/Test/Unit/Console/CommandLoader/AggregateTest.php:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * @dataProvider provideTestCasesForGet
+     */
+    public function testGet($firstCmd, $secondCmd)
```
Introduce strict types.

lbajsarowicz reviewed (commented):
Just sharing my thoughts seeing your work :)

ihor-sviziev commented:
@magento run all tests

gabrieldagama commented on lib/internal/Magento/Framework/Console/Cli.php:
```diff
@@ -101,6 +103,16 @@ public function __construct($name = 'UNKNOWN', $version = 'UNKNOWN')
+
+        $this->setCommandLoader($this->objectManager->create(Aggregate::class, [
+            'commandLoaders' => $commandLoaders
+        ]));
```
I would extract this to a function to make it easier to understand, the construct is already quite big, by adding just a function call there it would be cleaner.

gabrieldagama commented on setup/src/Magento/Setup/Console/CommandLoader.php:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class CommandLoader implements CommandLoaderInterface
```
By introducing this `CommandLoader`, can't we remove the `Magento\Setup\Console\CommandList`?

mattwellss replied:
I believe it can be removed... My goal was to minimize potential BC breaks by leaving the class in place, but since I'm not initializing it in the CLI (https://github.com/magento/magento2/pull/29355/files/2895c6e436fb42480f8f50795066e85af1da2f51#diff-4eb287b4e24c054a400e5f7b40dab0cbL146-L149), it wouldn't be *true* backwards compatibility.

mattwellss replied:
Resolved in 2d99944

gabrieldagama replied:
I guess this should be fine, the developers always depend on `CommandListInterface`, also the Setup CommandList doesn't accept any command via dependency injection.

gabrieldagama reviewed (changes requested):
Hi @mattwellss, thanks for the pull request, it looks great! Please follow some small comments on it.

I can definitely see the benefit of using `CommandLoader` over `CommandList`, by deferring the creation of the command objects just when we actually call it. From the performance perspective, I'm not having that much of a difference, on my tests, I'm clearing the cache on both runs to make sure the performance is actually increased.

So when changing branches I'm actually running:
```bin/magento cache:clean && time bin/magento setup:install --help```

Running on 2.4-develop:
```bin/magento setup:install --help  13.61s user 0.85s system 91% cpu 15.854 total```

Running on implement-command-loader:
```bin/magento setup:install --help  13.38s user 0.79s system 92% cpu 15.309 total```

Could you confirm if the performance increase is similar to what I'm getting here?

Let me know your thoughts! Thanks!

mattwellss commented:
@gabrieldagama, thanks for the review. I appreciate it!

Regarding performance:
1. The biggest improvement will be noticed when running a common command rather than the two which cause the biggest performance drain, `cache:clean` is a good example
   _This is because my intent is to improve performance when commands such as `setup:install` don't need to be initialized. If the `setup:install` command is initialized, the performance benefit of the command loader is partially lost_
1. The improvements made here are largely overshadowed by DI initialization (and all the other uncached stuff that has to happen when the cache is empty), so my goal is to improve performance while the cache is warm

Some results:

On _2.4-develop_:
```
./bin/magento cache:clean  0.70s user 0.16s system 98% cpu 0.881 total
```

On _implement-command-loader_:
```
./bin/magento cache:clean  0.58s user 0.15s system 97% cpu 0.752 total
```

Let me know if you see similar percentage improvement (the command loader version finishes in about 85% of the time of the base branch).

On my installation of Magento, there are 57 commands added via the `Magento\Framework\Console\CommandList`. All of these can be migrated to use the CommandLoader if we want to squeeze a bit more performance benefit out of the change. I'm not sure if we'll see much improvement, though... when I profiled a `./bin/magento list` command, only the `setup:install` and `setup:config:set` commands stood out as particularly slow to initialize.

gabrieldagama commented:
Thanks for the detailed reply @mattwellss! I understand your point, I will do a further investigation here, but on my environment, with warmed cache, there is not much difference in performance.

Regards the other using `CommandLoader` for the other commands, it may be backward incompatible, currently, developers rely on `CommandListInterface` to add new commands, so any change there will be backward-incompatible (https://devdocs.magento.com/guides/v2.4/extension-dev-guide/cli-cmds/cli-howto.html). That doesn't mean it is not possible, but we may need to be a little bit more careful if we do it, but as you said, if there are no real performance improvements, it may not be necessary.

Thanks!

gabrieldagama commented:
@magento run all tests

mattwellss commented:
@gabrieldagama regarding:
> Regards the other using CommandLoader for the other commands, it may be backward incompatible, currently, developers rely on CommandListInterface to add new commands, so any change there will be backward-incompatible (https://devdocs.magento.com/guides/v2.4/extension-dev-guide/cli-cmds/cli-howto.html). That doesn't mean it is not possible, but we may need to be a little bit more careful if we do it, but as you said, if there are no real performance improvements, it may not be necessary.

That's a great point about BC and also documentation. Here's how I'd work around it:
* `@deprecate` the `CommandListInterface` and `CommandList` in framework code
* Migrate commands which rely on the command list to use command loader
* Leave the command list initialization in place (`\Magento\Framework\Console\Cli::getApplicationCommands`, `\Magento\Framework\Console\Cli::getVendorCommands`) for custom and extension code that relies on it
* Update documentation to recommend use of CommandLoader

That approach **a)** won't break any existing code, **b)** helps to "document by example" (in addition to regular documentation),  **c)** gains whatever performance improvement is available.

Thanks again for reviewing. Let me know what you think.

gabrieldagama commented:
Hi @mattwellss, agree with you on steps there to introduce `CommandLoader` to general commands (outside Setup module).  I guess we can create a separate issue for it and work this out on separate PR (once this gets merged), what do you think?

There are some small fixes that are required from the `Static Tests`, are you able to address them? The functional failures aren't related to this PR.

Thanks for your contribution!

mattwellss commented:
> I guess we can create a separate issue for it and work this out on separate PR (once this gets merged), what do you think?

Good plan

Regarding static test failures, I'm not sure what to do about this error:
> Data set: lib/internal/Magento/Framework/Console/Cli.php
Cli.php depends for non-library classes Magento\Setup\Console\CommandLoader Magento\Setup\Console\CommandLoader Magento\Setup\Console\CommandLoader

It appears to be repeated a large number of times in [testCheckDependencies](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/d5b425a3adb3f2d714fd108e106fe5a4/Statics/allure-report-ce/index.html#suites/6ac5b45bf194ca9e37478df91bf3abac/95370e5c379a01ec/). I believe the error should've already existed for `Magento\Setup\ConsoleCommandList`'s usage inside `Cli` as well. I made sure to check `class_exists` before creating a new instance, so even if Magento Setup code is removed, the Cli should still run.

mattwellss commented:
@magento run all tests

engcom-Echo commented:
Hi, @mattwellss are you going to continue progress on that? Thanks!

mattwellss commented:
Hey, sorry about the delay. I got busy with other work. I fixed the errors in [`Magento\Test\Php\LiveCodeTest`](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/bc59967a3e931f8357ab0caca45634e5/Statics/allure-report-b2b/index.html#suites/51d3ab4d5d77a0bc6a7ea344e9e7e6be), aside from those in `*Command` classes (which got caught up in the validation because I modified them). The errors in those classes, such as phpcs direct throw warnings in the `AdminUserCreateCommand`, predated these changes, so I didn't attempt to fix them. Otherwise, the code quality checks should be fine. If this is fine, please request a test run and review the code.

mattwellss commented:
@engcom-Echo, please note my comment from a couple weeks ago.

ihor-sviziev reviewed (changes requested):
Hi @mattwellss,
Now the 2.5-develop branch was already created. 🎉

Could you re-target this PR to the 2.5-develop branch and resolve merge conflicts, so we'll be able to move it forward?
Thank you

ihor-sviziev commented on lib/internal/Magento/Framework/Console/Cli.php:
```diff
@@ -256,4 +254,23 @@ private function documentRootResolver(array $config = []): array
+    private function getCommandLoader(): Console\CommandLoader\CommandLoaderInterface
+    {
+        $commandLoaders = [];
+        if (class_exists(SetupCommandLoader::class)) {
```
Could you explain why this class might not exist?

mattwellss replied:
My goal was to avoid regressions, so I tried to translate this class literally to mirror how `CommandList` classes were used. Here's what inspired the check: https://github.com/magento/magento2/blob/1e846a4aa38f1ff5d7238d70df9d887b3bdf5acb/lib/internal/Magento/Framework/Console/Cli.php#L146

I'm not sure if any likely scenario exists where the `Magento\Setup` classes aren't available, but that's presumably the hedge this check is betting against. Let me know how you'd like to move forward... I'm happy to remove the check if we don't need to worry about cases where users of Magento have excluded "setup" classes from autoloading.

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader/Aggregate.php:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * @param array $commandLoaders
+     */
+    public function __construct($commandLoaders = [])
```
```suggestion
    public function __construct(array $commandLoaders = [])
```

ihor-sviziev commented on setup/src/Magento/Setup/Console/CommandLoader.php:
```diff
@@ -0,0 +1,116 @@
+     * @param string $name
+     * @return SymfonyCommand
+     */
+    public function get($name)
```
```suggestion
    public function get($name): SymfonyCommand 
```

ihor-sviziev commented on setup/src/Magento/Setup/Console/CommandLoader.php:
```diff
@@ -0,0 +1,116 @@
+     * Generate a symfony console command with the laminas service manager
+     *
+     * @param string $name
+     * @return SymfonyCommand
```
```suggestion
     * @return SymfonyCommand
     * @throws CommandNotFoundException
```

ihor-sviziev commented on setup/src/Magento/Setup/Console/CommandLoader.php:
```diff
@@ -0,0 +1,116 @@
+     * @param string $name
+     * @return bool
+     */
+    public function has($name)
```
```suggestion
    public function has($name): bool
```

ihor-sviziev commented on setup/src/Magento/Setup/Console/CommandLoader.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return string[]
+     */
+    public function getNames()
```
```suggestion
    public function getNames(): array
```

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader.php:
```diff
@@ -0,0 +1,77 @@
+     * @return \Symfony\Component\Console\Command\Command
+     * @throws CommandNotFoundException
+     */
+    public function get($name)
```
```suggestion
    public function get($name): \Symfony\Component\Console\Command\Command
```

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader.php:
```diff
@@ -0,0 +1,77 @@
+     * @param string $name
+     * @return bool
+     */
+    public function has($name)
```
```suggestion
    public function has($name): bool
```

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader.php:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return string[]
+     */
+    public function getNames()
```
```suggestion
    public function getNames(): array
```

ihor-sviziev commented on lib/internal/Magento/Framework/Console/Cli.php:
```diff
@@ -256,4 +254,23 @@ private function documentRootResolver(array $config = []): array
+        if (class_exists(SetupCommandLoader::class)) {
+            $commandLoaders[] = new SetupCommandLoader($this->serviceManager);
+        }
+        $commandLoaders[] = $this->objectManager->create(CommandLoader::class);
```
Can we define a list of these loaders via di.xml, so custom modules could also define their own loaders if needed?

ihor-sviziev replied:
It was done in the CommandLoader.

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader/Aggregate.php:
```diff
@@ -0,0 +1,78 @@
+     * @param string $name
+     * @return bool
+     */
+    public function has($name)
```
```suggestion
    public function has($name): bool
```

ihor-sviziev commented on lib/internal/Magento/Framework/Console/CommandLoader/Aggregate.php:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return string[]
+     */
+    public function getNames()
```
```suggestion
    public function getNames(): array
```

ihor-sviziev reviewed (changes requested):
I just looked at this PR one more time - I have a few suggestions. Please review them and update your PR accordingly.

mattwellss commented:
@ihor-sviziev, thanks for giving this some attention! I'll take a look at your feedback in the next few days and change the target to 2.5-dev.

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
Hello @gabrieldagama could you please help to understand why test builds fall without reports?
Thank you.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@gabrieldagama @sidolov could you help to understand why tests are failing with not build reports?

mattwellss committed:
Implement command loader

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
In the integration test results I see the following error:
```
Exception: PHP Fatal error:  Uncaught Error: Class 'Magento\Setup\Console\Command\MaintenanceAllowIpsCommand' not found in /var/www/html/lib/internal/Magento/Framework/Console/Cli.php:241
```
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/28f78a38ad38a8da83a25123b1bea786/Integration/console-error-logs.html
Could you update your PR in order to fix that issue?

mattwellss commented:
@ihor-sviziev, looks like that command was moved from `Magento_Setup` to `Magento_Backend`. 

https://github.com/magento/magento2/commit/4e40bc152a530d0bf549adb36817b1d5fced9e8b#diff-8c9bf1dd971391daf0211e3b5c134f73ab4a0bf76a78eda7c7e6e0afe2190d7e

Because (at the moment) this PR's scope is limited to updating `Magento_Setup` to use command loader, I simply removed the references to the "maintenance" commands from `\Magento\Setup\Console\CommandLoader::$commands`

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests B2B

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8747](https://jira.corp.magento.com/browse/ENGCOM-8747) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

magento-engcom-team commented:
@mattwellss thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

ihor-sviziev commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
@magento run Functional Tests B2B

engcom-Lima commented:
@magento run all tests

ihor-sviziev commented:
@engcom-Lima, maybe we can move the changes  on top of 2.4-develop branch and move it forward?

engcom-Lima commented:
As this PR is targeted to 2.5-develop the other PR takes the precedence.

ihor-sviziev commented:
FYI I'll try moving this PR to 2.4-delelop branch.
The original history I left there:
https://github.com/ihor-sviziev/magento2/compare/2.5-develop...ihor-sviziev:magento2:implement-command-loader

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Implement command loader

Fix test failures

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
https://github.com/magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (approved):
Update: I squashed all changes to a single commit, rebased on top of the 2.4-develop branch + fixed static test failures. https://github.com/magento/magento2/pull/29355#issuecomment-1750844588

✔ Approved from my side.

Failing tests look not related to changes from this PR.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (approved):
Update: I squashed all changes to a single commit, rebased on top of the 2.4-develop branch + fixed static test failures. https://github.com/magento/magento2/pull/29355#issuecomment-1750844588

✔ Approved from my side.

Failing tests look not related to changes from this PR.

ihor-sviziev committed:
Merge branch '2.4-develop' into implement-command-loader

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Implement command loader

Fix test failures

ihor-sviziev commented:
@magento run Unit Tests

engcom-Echo committed:
Merge branch '2.4-develop' into implement-command-loader

engcom-Echo commented:
@magento run all tests

engcom-Echo commented:
@mattwellss Thank you for the contribution.

We have followed below steps for Verification.
 
1. Install clean/fresh Magento Instance.
2. Checkout to 2.4-develop branch.
3. Run a console command through time; for exmaple: time ./bin/magento setup:install --help and note how long it takes

<img width="644" alt="Screenshot 2024-05-28 at 1 33 32 PM" src="https://github.com/magento/magento2/assets/51681435/23a9333f-1d6c-4f4f-be7f-283f7b9112bf">

5. Check out this branch locally (mattwellss:implement-command-loader)
6. Run the same console command again, and note that it is not faster than before please check below screenshot.

<img width="645" alt="Screenshot 2024-05-28 at 1 33 19 PM" src="https://github.com/magento/magento2/assets/51681435/7e906887-aa27-4460-9c20-3a513d86de84">

Can you please suggest if we have missed anything here.

engcom-Echo commented:
@mattwellss did you get time to check [comment](https://github.com/magento/magento2/pull/29355#issuecomment-2134929271).
For now we are moving this PR to `On Hold`

mattwellss commented:
In a comment almost four years ago I mentioned that `setup:install` is a poor verification example: https://github.com/magento/magento2/pull/29355#issuecomment-679412344. I shouldn't have used it as the test subject for verification.

Regardless, it's pretty shocking that this PR's run of `setup:install --help` would take over 100s. Clearly there's something wrong, but I'm unable to reproduce your experience:
<img width="726" alt="image" src="https://github.com/magento/magento2/assets/1776257/31d6744c-cc85-433c-8777-589cadeb0ab4">

As you can see from my screenshot, this PR's code displays `--help` a little faster. There's variation run-to-run, but it seems consistently a bit faster.

A final note about performance expectations: these changes cannot cause a massive performance boost, and they weren't designed to do so. Rather, they lay the groundwork for future improvement by providing a framework for adding a command to the application without initializing it.

engcom-Echo commented:
@mattwellss We have tried to verify the PR changes based on [comment](https://github.com/magento/magento2/pull/29355#issuecomment-679412344) as well.
But we are not seeing any improvement.
Please see below screenshot for reference.
<img width="780" alt="Screenshot 2024-07-03 at 11 42 10 AM" src="https://github.com/magento/magento2/assets/51681435/184ca494-3e7d-4426-9fa7-b8d5642c95ac">

let us know if we have missed anything here.

ihor-sviziev commented:
@engcom-Echo, the issue is NOT reproducing on a clean magento installation, but it is reproducing when you have many different extensions from different vendors.

The technically difference is following _when you're running some specific command_:
- before changes: 
    - **magento loads all the available commands** and all their dependencies
    - then run a single command
- after changes:
    - **magento loads a single command** and it's dependencies
    - then tun a single command 

Results will highly depend on the amount of dependencies, quality of extensions, etc.

engcom-Charlie commented:
Hi @mattwellss,

Thank you for your contribution!

I have tested this PR only on core Magento instance as well as with extension.

1. Below is the result of **`time ./bin/magento cache:clean --help`** command on core Magento and PR instance

```
OUTPUT on Magento 2.4 develop 
./bin/magento cache:clean --help  0.60s user 0.22s system 63% cpu 1.294 total

OUTPUT on PR branch
./bin/magento cache:clean --help  3.81s user 0.21s system 99% cpu 4.050 total
```


2. Below is the result of **`time ./bin/magento cache:clean --help`** command on core Magento with Extension

```
OUTPUT on Magento 2.4 develop 
./bin/magento cache:clean --help  0.20s user 0.15s system 47% cpu 0.734 total

OUTPUT on PR branch
./bin/magento cache:clean --help  0.18s user 0.12s system 45% cpu 0.654 total
```


As per above result in #1 and #2, without any extension, the result is good in Magento 2.4-develop instance but with extensions we are getting good result in PR branch. Hence can you please confirm on this point and let me know if any other thing we need to take care here related to reproducing the issue or PR verification.

Thank you!

engcom-Charlie committed:
Merge branch '2.4-develop' into implement-command-loader

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Semantic Version Checker, Static Tests, Unit Tests

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Unit Tests

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

hostep commented:
> ```
> OUTPUT on PR branch
> ./bin/magento cache:clean --help  3.81s user 0.21s system 99% cpu 4.050 total
> ```

There is no way this output is accurate. Probably your computer was working hard on something else while you tested it, which caused the result to take a lot longer then it should?

Best is probably to run your tests up to 5 or 10 different times, drop the highest and lowest value and make an average of the other runs.

engcom-Charlie commented:
> > ```
> > OUTPUT on PR branch
> > ./bin/magento cache:clean --help  3.81s user 0.21s system 99% cpu 4.050 total
> > ```
> 
> There is no way this output is accurate. Probably your computer was working hard on something else while you tested it, which caused the result to take a lot longer then it should?
> 
> Best is probably to run your tests up to 5 or 10 different times, drop the highest and lowest value and make an average of the other runs.

Hi @hostep,

As you see above comment threads, many times we have tested this and conveyed the same here. This above result also posted after many execution.

Thank you!

engcom-Charlie commented:
As mentioned [here](https://github.com/magento/magento2/pull/29355#issuecomment-2273497399), moving this PR to On Hold until confirmation.

engcom-Charlie commented:
We have tested the PR once again and the results are as below. On PR branch the initialization time is reduces compared to Magento 2.4 develop hence moving this PR to Extended Testing now.

**On Magento 2 without extension**

<img width="1420" alt="Screenshot 2024-08-09 at 16 56 00" src="https://github.com/user-attachments/assets/497fa593-65da-4e90-8060-a78d76404143">

**On PR branch without extension**

<img width="1420" alt="Screenshot 2024-08-09 at 16 56 24" src="https://github.com/user-attachments/assets/80967b0a-5839-4695-867c-7de9996935df">

**On Magento 2 with extension**

<img width="1420" alt="Screenshot 2024-08-09 at 16 56 57" src="https://github.com/user-attachments/assets/945a5a98-8194-4e4b-b673-83b7d1d875bd">


**On PR branch with extension**

<img width="1420" alt="Screenshot 2024-08-09 at 16 57 38" src="https://github.com/user-attachments/assets/0ab1d9a3-0600-4df1-ba7a-a3bf76cbb4e2">

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

engcom-Charlie commented:
Failures of Functional CE, B2B and EE are known open issues related to Page builder repo. Hence moving this to Merge in Progress.

**Functional B2B:**
Run 1:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/70f9699d843952565c1e8c59d444ebfd/Functional/allure-report-b2b/index.html#categories

Run 2: 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/84de6348f4db860c0062029dce8d20b2/Functional/allure-report-b2b/index.html#categories

**Functional CE:**
**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/b230b1032ac0fde0f2e1372620f8d865/Functional/allure-report-ce/index.html#categories

**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/44542d5543464fba358b48be11675467/Functional/allure-report-ce/index.html#categories


**Functional EE:**
**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/c74867cac5453927346efee29dde3ec5/Functional/allure-report-ee/index.html

**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/29355/4ecca2a37665531176a17f2adf83ba56/Functional/allure-report-ee/index.html#categories

**Known JIRA:**
[ACQE-6695](https://jira.corp.adobe.com/browse/ACQE-6695)

engcom-Dash commented:
Moving this PR to extended testing since the copyright tag has not been updated in the mentioned files.

engcom-Dash commented:
@magento run Static Tests

engcom-Dash committed:
29355: Fix static failure related to copyright tag

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
Got a green build so moving it to merge in progress.
<img width="1790" alt="image" src="https://github.com/user-attachments/assets/572d753b-9355-43ec-a24d-cfd51c2430a6">
